blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,784 @@
+
+You'll also notice one example where the new method is *less specific*.
+It is not clear why such methods should be invalidating, and this may be a Julia bug.
+
+### Partial specialization
+
+If you try
+
+```julia-repl
+julia> trees = invalidation_trees(@snoopr using StaticArrays)
+```
+
+you'll see a much longer output.
+A large number of invalidations derive from the fact that StaticArrays defines a method for `!=`, which invalidates the fallback definition
+
+```
+!=(x, y) = !(x == y)
+```
+
+Since such definitions account for hundreds of nominal invalidations, it would be well worth considering whether it is possible to delete the custom `!=` methods.
+For example, if they are purely for internal use you could modify each caller to
+use the default method.
+
+The vast majority of the rest appear to derive from ambiguities.
+However, one more interesting case we've not seen before is
+
+```julia-repl
+julia> tree = trees[end-7]
+insert unsafe_convert(::Type{Ptr{T}}, m::Base.RefValue{FA}) where {N, T, D, FA<:FieldArray{N,T,D}} in StaticArrays at /home/tim/.julia/packages/StaticArrays/mlIi1/src/FieldArray.jl:124 invalidated:
+   mt_backedges: signature Tuple{typeof(Base.unsafe_convert),Type{Ptr{_A}} where _A,Base.RefValue{_A} where _A} triggered MethodInstance for unsafe_convert(::Type{Ptr{Nothing}}, ::Base.RefValue{_A} where _A) (159 children) more specific
+```
+
+In this case, the signature that triggered the invalidation, `Base.unsafe_convert(::Type{Ptr{_A}} where _A, ::Base.RefValue{_A} where _A)`,
+has been only partially specified: it depends on a type parameter `_A`.
+Where does such a signature come from?
+You can extract this line with
+
+```julia-repl
+julia> trigger = tree[:mt_backedges, 1]
+MethodInstance for unsafe_convert(::Type{Ptr{Nothing}}, ::Base.RefValue{_A} where _A) at depth 0 with 159 children
+
+julia> trigger.children
+5-element Array{SnoopCompile.InstanceTree,1}:
+ MethodInstance for unsafe_convert(::Type{Ptr{Nothing}}, ::Base.RefValue{_A} where _A) at depth 1 with 0 children
+ MethodInstance for unsafe_convert(::Type{Ptr{T}}, ::Base.RefValue{Tuple{Vararg{T,N}}}) where {N, T} at depth 1 with 2 children
+ MethodInstance for _show_default(::Base.GenericIOBuffer{Array{UInt8,1}}, ::Any) at depth 1 with 113 children
+ MethodInstance for _show_default(::IOContext{Base.GenericIOBuffer{Array{UInt8,1}}}, ::Any) at depth 1 with 37 children
+ MethodInstance for _show_default(::IOContext{REPL.Terminals.TTYTerminal}, ::Any) at depth 1 with 2 children
+```
+
+and see all the `MethodInstance`s that called this one.
+You'll notice three `_show_default` `MethodInstance`s here;
+a little digging reveals that this is defined as
+
+```
+function _show_default(io::IO, @nospecialize(x))
+    t = typeof(x)
+    ...
+end
+```
+
+So the `@nospecialize` annotation, designed to reduce the number of cases when `_show_default` needs to be recompiled, causes the methods *it* uses to become triggers for invalidation.
```

Comment:
Perhaps I'm dense but it wasn't immediately obvious to me why this was the case.  Speaking out loud, is this because within this method body, `x::Any`... and thus any time you `f(..., x, ...)` for `f`s with small method tables it'll inline and/or union-split?

How's this related to ambiguous cases?

Reply:
Yeah, this may be a problematic example. I'll try to find something better.